src/function/mtk.jl

```diff
@@ -7,101 +7,31 @@ AutoModelingToolkit() = AutoModelingToolkit(false, false)
 
 function instantiate_function(f, x, adtype::AutoModelingToolkit, p, num_cons = 0)
     p = isnothing(p) ? SciMLBase.NullParameters() : p
-    sys = ModelingToolkit.modelingtoolkitize(OptimizationProblem(f, x, p))
 
-    hess_prototype, cons_jac_prototype, cons_hess_prototype = nothing, nothing, nothing
+    sys = ModelingToolkit.modelingtoolkitize(OptimizationProblem(f, x, p), num_cons = num_cons)
+    sys = ModelingToolkit.structural_simplify(sys)
+    f = OptimizationProblem(sys, x, p, grad = true, hess = true, obj_sparse = adtype.obj_sparse, cons_j = true, cons_h = true, cons_sparse = adtype.cons_sparse).f
```

Comment:
This is potentially very costly. Shouldn't it be an option. Or even better if solve is called for a solver that needs gradient's and Hessians it should be generated, otherwise not.

Reply:
I agree it will potentially be costly. But for now, it's better to maintain uniformity with the other wrappers (they populate everything by default), we will move to the check and populate very soon but it would be better to do it in a self-contained PR